File: Chem 272 Project 2.cpp
#include <vector>
#include <iostream>
#include <random>
#include <cmath>
#include "matplotlibcpp.h"

namespace plt = matplotlibcpp;

std::vector<std::vector<double>> CoordsToPotential(const std::vector<double>& Xinit, const std::vector<double>& Yinit, double a, double b, int Particles) {

    std::vector<std::vector<double>> Phi(Particles, std::vector<double>(Particles, 0.0));
    
    double sigma = std::pow((a/b), (1.0/6.0));
    double r_cutoff = 3.0 * sigma;
    double Zero_equiv = std::pow(10, -16)

    for (int i = 0; i < Particles; ++i) {
        for (int j = 0; j < Particles; ++j) {
                Phi[i][j] = Zero_equiv;
            }
        }

    for (int i = 0; i < Particles; ++i) {
        for (int j = i+1; j < Particles; ++j) {

            double Dx = Xinit[i] - Xinit[j];
            double Dy = Yinit[i] - Yinit[j];
            double r = std::sqrt(Dx*Dx + Dy*Dy);

            if (r > r_cutoff) {
                Phi[i][j] = Phi[j][i] = Zero_equiv;
                continue;
            }

            double Utot = (a / std::pow(r, 12)) - (b / std::pow(r, 6));
            Phi[i][j] = Phi[j][i] = Utot;
            }
        }
    return Phi;
}

std::vector<double> Utot_Calc(const std::vector<std::vector<double>>& Phi) {
    int Particles = Phi.size();
    std::vector<double> Utot(Particles, 0.0);
    
    for (int i = 0; i < Particles; ++i) {
        for (int j = 0; j < Particles; ++j) {
            Utot[j] += Phi[i][j];
        }
    }
    return Utot;
}

int main() {

int Niter = 10000;
int Num_Particles = 1000;
double a = 0.01;
double b = 0.01;
double k = 0.2;
double Temp = 298.15;
double TempScale = k*Temp

std::random_device rd; // Provides a non-deterministic seed
std::mt19937 gen(rd()); // Mersenne Twister engine seeded by random_devic

std::vector<double> Xinit(Num_Particles, 0);
std::vector<double> Yinit(Num_Particles, 0);
std::vector<double> Mass(Num_Particles, 0);
std::vector<double> No_BM_Probz(Num_Particles, 0);

std::uniform_real_distribution<double> positions(-100.0, 100.0);
std::uniform_real_distribution<double> distru(0.0, 1.0);
std::uniform_real_distribution<double> mass_val(1.0, 20.0);
std::uniform_real_distribution<double> dist_move(-1.0, 1.0);

for (auto &x : Xinit) {
    x = positions(gen);
}
   
for (auto &y : Yinit) {
    y = positions(gen);
}

for (auto &m : Mass) {
    m = mass_val(gen);
}

std::vector<double> Mass_Effect(Num_Particles, 0.0);
for (int i = 0; i < Num_Particles; ++i) {
        Mass_Effect[i] = 1.0 / Mass[i];
}

std::vector<std::vector<double>> Phi = CoordsToPotential(Xinit, Yinit, a, b, Num_Particles);
std::vector<double> Utot = Utot_Calc(Phi);

std::vector<double> dx_move(Num_Particles, 0.0);
std::vector<double> dy_move(Num_Particles, 0.0);
std::vector<double> Xinit_new(Num_Particles, 0.0);
std::vector<double> Yinit_new(Num_Particles, 0.0);
std::vector<std::vector<double>> Phi_new(Num_Particles, std::vector<double>(Particles, 0.0));
std::vector<double> Utot_new (Num_Particles, 0.0);
std::vector<double> Utot_diff(Num_Particles, 0.0);
std::vector<double> Boltzmann_Dist(Num_Particles, 0.0);

for (int n = 0; n < Niter; ++n){
    for (int i = 0; i < Num_Particles; ++i) {
        dx_move[i] = dist_move(gen)*Mass_Effect[i];
        dy_move[i] = dist_move(gen)*Mass_Effect[i];
    }

    for (int i = 0; i < Num_Particles; ++i) {
         Xinit_new[i] = Xinit[i] + dx_move[i];
         Yinit_new[i] = Yinit[i] + dy_move[i];
    }

    Phi_new = CoordsToPotential(Xinit_new, Yinit_new, a, b, Num_Particles);
    Utot_new = Utot_Calc(Phi_new);

    for (int i = 0; i < Num_Particles; ++i) {
        Utot_diff[i] = Utot[i] - Utot_new[i];
        Boltzmann_Factor[i] = std::exp(-1 * Utot_diff[i]/ TempScale);
    }

    for (auto &p : No_BM_Probz){
        p = distru(gen);
        }

    for (size_t i=0; i < Num_Particles; ++i) {
        if (Utot_diff[i]>0) {
            Xinit[i] = Xinit_new[i];
            Yinit[i] = Yinit_new[i];
        }
        else if (Boltzmann_Factor[i] > No_BM_Probz[i] && Utot_diff[i]<0)
            Xinit[i] = Xinit_new[i];
            Yinit[i] = Yinit_new[i];
        }

    Phi = CoordsToPotential(Xinit, Yinit, a, b, Num_Particles);
    Utot = Utot_Calc(Phi);

    if (n%1000 == 0) {
        plt::title("Particle Movement");
        plt::scatter(Xinit, Yinit);
        plt::xlim(-100, 100);
        plt::ylim(-100, 100);
        plt::show();
    }
    }
    return 0;
}

File: Chem 272 Project 2.py
import numpy as np
import matplotlib.pyplot as plt

def Plot_Particles(Xinit, Yinit, Niter, mass):
    plt.title(f'After N = {Niter} iterations')
    plt.scatter(Xinit, Yinit, marker='o', s=mass, facecolors='grey', edgecolors='black', alpha=0.5)
    plt.xlim(-100, 100)
    plt.ylim(-100, 100)
    plt.show()

def CoordsToPotential(Xinit, Yinit, a, b, Particles):
    
    X = np.tile(Xinit, (Particles, 1)) #tile 1D array of x coords into matrix
    Y = np.tile(Yinit, (Particles, 1)) #tile 1D array of y coords into matrix
    
    Dx = X - X.transpose() #matrix of distances between xi and xj
    Dy = Y - Y.transpose() #matrix of distances between yi and yj

    Eye = np.eye(Particles) #square matrix with diagonal of 1s with size of Particles
    Ones = np.ones((Particles,)) #1D array of 1s

    r = np.sqrt((Dx**2) + (Dy**2)) #matrix of distances between particles
    r += Eye*(10**16) #diagonal of large infinitely large
    Phi = (a / r**12) - (b / r**6) #calculate potential with lennard jones

    Utot = np.dot(Phi, Ones) #sum of potential each particle feels into one row
    
    return Utot

def Utot_Move_Particle(Xinit, Yinit, Niter, Particles, a, b, k, Temp, mass):
    
    Utot = CoordsToPotential(Xinit, Yinit, a, b, Particles) #starting potential

    for n in range(Niter):
        
        Mass_Effect = 1/mass #----------------------------
        dx = Mass_Effect * np.random.choice([-1, 1], (Particles,)) #choose -1 OR 1 --> 1D array size of Particles for dx
        dy = Mass_Effect * np.random.choice([-1, 1], (Particles,)) #choose -1 OR 1 --> 1D array size of Particles for dy

        Xinit_new = Xinit + dx #move x direction
        Yinit_new = Yinit + dy #move y direction

        Utotnew = CoordsToPotential(Xinit_new, Yinit_new, a, b, Particles) #potential after move
        Utotdiff = Utotnew - Utot

        Utot_change_index = np.argwhere(Utotdiff < 0).flatten() #indices where potential decreases
        
        Xinit[Utot_change_index] = Xinit_new[Utot_change_index] #move particles (by taking indices where potential decreases) to new x coord
        Yinit[Utot_change_index] = Yinit_new[Utot_change_index] #move particles (by taking indices where potential decreases) to new x coord

        Prob_No_BM = np.random.rand(Particles) #Probability no brownian motion occurs
            
        Boltzmann_Dist = np.exp(-Utotdiff/(k*Temp)) #probability of particle moving
        Brown_Index = np.argwhere((Utotdiff > 0) & (Boltzmann_Dist > Prob_No_BM)).flatten() #particles at indices where potential increases AND boltz prob > probability no motion

        Xinit[Brown_Index] = Xinit_new[Brown_Index] #particles who's energy will increase can move if boltzman probability is high
        Yinit[Brown_Index] = Yinit_new[Brown_Index] 
        
        Utot = CoordsToPotential(Xinit, Yinit, a, b, Particles) #recalculate Utot


        if n % 100 == 0:
            Plot_Particles(Xinit, Yinit, n, mass)

    return Xinit, Yinit, Utot

class ParticleMotion:

    def __init__(self, Niter=10000, Particles=1000, a=0.01, b=0.01, Pmin=-100, Pmax= 100, k=0.2, Temp=298.15):
        self.Niter = Niter
        self.Particles = Particles
        self.a = a
        self.b = b
        self.Pmin = Pmin
        self.Pmax = Pmax
        self.k = k
        self.Temp = Temp

        self.Xinit = np.random.uniform(Pmin, Pmax, (Particles,)) #initialize x coordinates of particles (1D np array)
        self.Yinit = np.random.uniform(Pmin, Pmax, (Particles,)) #initialize y coordinates of particles (1D np array)
        self.mass = np.random.uniform(1, 20, (Particles,)) #initialize random masses of particles (1D np array)

    def Run_Particle_Movement(self):
        Utot_Move_Particle(self.Xinit, self.Yinit, self.Niter, self.Particles, self.a, self.b, self.k, self.Temp, self.mass)

ParticleMotion().Run_Particle_Movement()
